File: 5-Problem-A-Day/251. Flatten 2D Vector.java
class Vector2D {

    int i, j;
    int[][] v;
    public Vector2D(int[][] v) {
        this.v = v;
        i = 0;
        j = 0;
    }
    
    public int next() {
        hasNext();
        return v[i][j++];
    }
    
    public boolean hasNext() {
        while (i != v.length && j == v[i].length) {
            j = 0;
            i++;
        }
        return i != v.length;
    }
}

/**
 * Your Vector2D object will be instantiated and called as such:
 * Vector2D obj = new Vector2D(v);
 * int param_1 = obj.next();
 * boolean param_2 = obj.hasNext();
 */

File: 2021/251. Flatten 2D Vector.cpp
class Vector2D {
public:    
    int outIndex = 0, inIndex = 0;
    vector<vector<int>> v;
    Vector2D(vector<vector<int>>& v) {
        this->v = v;
    }
    
    int next() {
        if (!hasNext()) return -1;
        inIndex++;
        return v[outIndex][inIndex - 1];
    }
    
    bool hasNext() {
        while (outIndex < v.size() && inIndex >= v[outIndex].size()) {
            outIndex++;
            inIndex = 0;
        }
        return outIndex < v.size();
    }
};

/**
 * Your Vector2D object will be instantiated and called as such:
 * Vector2D* obj = new Vector2D(v);
 * int param_1 = obj->next();
 * bool param_2 = obj->hasNext();
 */